### ISSUE TITLE
Using overlay as root filesystem systemd-machine-id-commit.service fails

### ISSUE THREAD

**Issue by djgera** [systemd/systemd#729]
This is booting Arch Linux with an empty /etc/machine-id file using overlay as root filesystem (squashfs-ro + tmpfs-rw) The tested since systemd (218-222), same thing happens.

Failed to unmount transient /etc/machine-id file in
our private namespace: Invalid argument

Just about this message, should care about it? /etc/machine-id gets writen with an id.

 # hostnamectl
    Static hostname: archiso
          Icon name: computer-vm
            Chassis: vm
         Machine ID: b0e5ffa9728f421dacf4564936dce6e1
            Boot ID: 92e4f6a2202f49798cb0e4782e791daf
     Virtualization: kvm
   Operating System: Arch Linux
             Kernel: Linux 4.1.2-2-ARCH
       Architecture: x86-64
 #

**Comment by poettering** [comment]
Hmm, overlayfs is a really fucked up file system, as a ton of operations don't work correctly on it.

And chance you can get an strace of this or so? (maybe boot via initrd, open a shell from the initrd on some high vt, then strace pid1 from there, and continue booting)

**Comment by djgera** [comment]
Yup, I can do, here is it [#1](too big for paste) ;)
Just attached to pid1 before doing a ""exec switch_root systemd"" from initrd. (pid1 is ash from busybox)

[#1] https://www.dropbox.com/s/3olpkd1awoyqbxq/strace-pid1.log.gz

**Comment by richardmaw-codethink** [comment]
The only EINVAL I see in that log is from attempting to clear the devices list for a populated cgroup, which explicitly ignores the EINVAL.

The actual error doesn't come from PID1, but rather from `systemd-machine-id-commit`, which is in `systemd-machine-id-commit.service`.

So we need to either have that strace'd instead (it should be easier to insert a /bin/strace before the ExecStart= in there than to strace pid 1), or to set the flag in strace to make it follow subprocesses.

**Comment by richardmaw-codethink** [comment]
At a wild guess: overlayfs thinks /etc/machine-id is a mount point because:
1.  It doesn't support `name_to_handle_at` and you don't have fdinfo, so it falls back to stat.
2.  overlayfs' stat passes through to the underlying backing file
3.  Because you made a change to `/etc/machine-id`, overlayfs made a copy in an upper layer, so there's a version of it on a different filesystem to that of `/etc/.`

Since it has all the properties of being on a different filesystem, and systemd assumes the only way that can happen (since it _is_ the only way it can happen on a POSIX compliant filesystem) is from a mount, systemd assumes it must be mounted, so attempts to unmount it, so it can replace its contents with a persistent machine-id.

**Comment by richardmaw-codethink** [comment]
@djgera As for whether you should worry about it, provided nothing is failing to start because `systemd-machine-id-commit.service` fails, you should be ok, as it fails before it can attempt to re-write your /etc/machine-id file.

**Comment by martinpitt** [comment]
See the thread starting on http://lists.freedesktop.org/archives/systemd-devel/2015-March/029163.html . Indeed the "is mountpoint" detection doesn't work for overlayfs, this would need a fallback in `path_is_mount_point()`.

**Comment by djgera** [comment]
OK, here is it.

```
execve("/usr/lib/systemd/systemd-machine-id-commit", ["/usr/lib/systemd/systemd-machine"...], [/* 2 vars */]) = 0
brk(0)                                  = 0x55e16aacb000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=36078, ...}) = 0
mmap(NULL, 36078, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fc102f8f000
close(3)                                = 0
open("/usr/lib/librt.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\220!\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=31672, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f8e000
mmap(NULL, 2128856, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fc102b6f000
mprotect(0x7fc102b76000, 2093056, PROT_NONE) = 0
mmap(0x7fc102d75000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6000) = 0x7fc102d75000
close(3)                                = 0
open("/usr/lib/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320`\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=142832, ...}) = 0
mmap(NULL, 2213040, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fc102952000
mprotect(0x7fc10296a000, 2093056, PROT_NONE) = 0
mmap(0x7fc102b69000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17000) = 0x7fc102b69000
mmap(0x7fc102b6b000, 13488, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fc102b6b000
close(3)                                = 0
open("/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260\10\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1979984, ...}) = 0
mmap(NULL, 3807760, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fc1025b0000
mprotect(0x7fc102749000, 2093056, PROT_NONE) = 0
mmap(0x7fc102948000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x198000) = 0x7fc102948000
mmap(0x7fc10294e000, 14864, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fc10294e000
close(3)                                = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f8d000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f8b000
arch_prctl(ARCH_SET_FS, 0x7fc102f8b740) = 0
mprotect(0x7fc102948000, 16384, PROT_READ) = 0
mprotect(0x7fc102b69000, 4096, PROT_READ) = 0
mprotect(0x7fc102d75000, 4096, PROT_READ) = 0
mprotect(0x55e168add000, 4096, PROT_READ) = 0
mprotect(0x7fc102f98000, 4096, PROT_READ) = 0
munmap(0x7fc102f8f000, 36078)           = 0
set_tid_address(0x7fc102f8ba10)         = 571
set_robust_list(0x7fc102f8ba20, 24)     = 0
rt_sigaction(SIGRTMIN, {0x7fc102957bb0, [], SA_RESTORER|SA_SIGINFO, 0x7fc102962660}, NULL, 8) = 0
rt_sigaction(SIGRT_1, {0x7fc102957c40, [], SA_RESTORER|SA_RESTART|SA_SIGINFO, 0x7fc102962660}, NULL, 8) = 0
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
brk(0)                                  = 0x55e16aacb000
brk(0x55e16aaec000)                     = 0x55e16aaec000
open("/proc/self/stat", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f97000
read(3, "571 (systemd-machine) R 569 569 "..., 1024) = 335
close(3)                                = 0
munmap(0x7fc102f97000, 4096)            = 0
access("/proc/vz", F_OK)                = -1 ENOENT (No such file or directory)
open("/run/systemd/container", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/proc/1/environ", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0400, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f97000
read(3, "TERM=linux\0", 1024)           = 11
read(3, "", 1024)                       = 0
close(3)                                = 0
munmap(0x7fc102f97000, 4096)            = 0
open("/proc/cmdline", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f97000
read(3, "BOOT_IMAGE=boot/x86_64/vmlinuz a"..., 1024) = 128
close(3)                                = 0
munmap(0x7fc102f97000, 4096)            = 0
access("/etc/initrd-release", F_OK)     = -1 ENOENT (No such file or directory)
ioctl(2, TCGETS, 0x7ffe227224c0)        = -1 ENOTTY (Inappropriate ioctl for device)
socket(PF_LOCAL, SOCK_DGRAM|SOCK_CLOEXEC, 0) = 3
getsockopt(3, SOL_SOCKET, SO_SNDBUF, [212992], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_SNDBUFFORCE, [8388608], 4) = 0
setsockopt(3, SOL_SOCKET, SO_SNDTIMEO, "\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
connect(3, {sa_family=AF_LOCAL, sun_path="/run/systemd/journal/socket"}, 29) = 0
openat(AT_FDCWD, "/etc", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC|O_PATH) = 4
name_to_handle_at(0x4, 0x55e168adaf93, 0x7ffe227221c0, 0x7ffe227221b8, 0) = -1 EOPNOTSUPP (Operation not supported)
name_to_handle_at(0x4, 0x55e168adafea, 0x7ffe22722250, 0x7ffe227221bc, 0x1000) = -1 EOPNOTSUPP (Operation not supported)
openat(4, "machine-id", O_RDONLY|O_NOCTTY|O_CLOEXEC|O_PATH) = 5
open("/proc/self/fdinfo/5", O_RDONLY|O_CLOEXEC) = 6
fstat(6, {st_mode=S_IFREG|0400, st_size=0, ...}) = 0
fstat(6, {st_mode=S_IFREG|0400, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f97000
read(6, "pos:\t0\nflags:\t012000000\nmnt_id:\t"..., 2048) = 35
read(6, "", 1024)                       = 0
read(6, "", 3072)                       = 0
close(6)                                = 0
munmap(0x7fc102f97000, 4096)            = 0
close(5)                                = 0
open("/proc/self/fdinfo/4", O_RDONLY|O_CLOEXEC) = 5
fstat(5, {st_mode=S_IFREG|0400, st_size=0, ...}) = 0
fstat(5, {st_mode=S_IFREG|0400, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc102f97000
read(5, "pos:\t0\nflags:\t012200000\nmnt_id:\t"..., 2048) = 35
read(5, "", 1024)                       = 0
read(5, "", 3072)                       = 0
close(5)                                = 0
munmap(0x7fc102f97000, 4096)            = 0
close(4)                                = 0
open("/etc/machine-id", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 4
read(4, "8b32a941581045699a8af1bda06ffe8c"..., 33) = 33
fstatfs(4, {f_type="TMPFS_MAGIC", f_bsize=4096, f_blocks=65536, f_bfree=63807, f_bavail=63807, f_files=127707, f_ffree=127640, f_fsid={0, 0}, f_namelen=255, f_frsize=4096}) = 0
close(4)                                = 0
open("/proc/self/ns/mnt", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 4
unshare(CLONE_NEWNS)                    = 0
mount(NULL, "/", NULL, MS_REC|MS_SLAVE, NULL) = 0
umount2("/etc/machine-id", 0)           = -1 EINVAL (Invalid argument)
sendmsg(3, {msg_name(0)=NULL, msg_iov(4)=[{"PRIORITY=3\nSYSLOG_FACILITY=3\nCOD"..., 166}, {"MESSAGE=", 8}, {"Failed to unmount transient /etc"..., 91}, {"\n", 1}], msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 266
close(4)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by richardmaw-codethink** [comment]
That's interesting, it looks like overlayfs also reporting different mnt_ids through `/proc/self/fdinfo`, since otherwise we'd see two calls to stat afterwards to check the root mountpoint case.

I'd hazard a guess as to this being because overlayfs' indirection leaves you a handle to the file as if you had opened it from the underlying filesystem directly, without any proxy to intercept the fdinfo check and report the mount ID of the proxied file.

Either way, we're sure this is because overlayfs is behaving un-posixy, and I agree with the suggestion on the mailing list that the best thing would be for overlayfs to be fixed, and that the best thing systemd should do in this case is degrade the unmount failure to be a warning that perhaps the mountpoint check was flawed.

This is not the only way the unmount could fail, as it could fail from a race where something else unmounts `/etc/machine-id` in the mean-time, but this could be mitigated by unsharing the mount namespace and turning off unmount propagation before checking the mount point, so it becomes impossible for it to be unmounted between checking and unmounting.

**Comment by poettering** [comment]
overlayfs semantics are really weird here. They basically confuses every software that wants to detect wether something is a mount point or not. I am pretty sure this is something to fix in overlayfs though, not in systemd. They need to somehow make sure the same mount id reported to userspace for all files of the same overlayfs mount. We shouldn't have to work-around this in userspace (in particular since I am not aware of any way we could...)

Please file a bug about this on kernel bugzilla, it needs to be fixed there, asking for a fix to /proc/fself/fdinfo/$FD based mnt_id reporting for files provided on overlayfs. Also, even better would be if name_to_handle_at() could also be used to determine the mnt_id correctly. Closing this here now.

**Comment by martinpitt** [comment]
Right, overlayfs exposes the fact whether a file/dir is in the overlay or underlay in the file system, i. e. in inodes, mount IDs/handles, etc. The only chance to see if a path is a real  mountpoint is to match it against `/proc/self/mounts`. There once was a patch to do this (http://lists.freedesktop.org/archives/systemd-devel/2015-March/029163.html), but it's indeed ugly and feels like a workaround. I dropped it from the Debian/Ubuntu packages too as it does not apply any more and we just don't care enough (this basically only affects the live system, and the failed machine-id-commit unit there doesn't hurt much).

**Comment by poettering** [comment]
I think passing through the dev_t from the underlying fs is probably acceptable, the way overlayfs currently does it. But passing through the mnt_id is certainly very wrong...

**Comment by richardmaw-codethink** [comment]
This appears to be a result of overlayfs doing the simple thing, and proxying the open by translating it to an open with a translated path, rather than opening the file and recording the path it was actually opened from. (see https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/fs/overlayfs/inode.c?id=v4.1#n339)
As a result, when we attempt to retrieve the `mnt_id` from `/proc/self/fdinfo`, it uses the translated path, which has a different mount id.

Linux v4.2 looks like it's changing the mechanism, so that the `f_path` structure points to the right thing though (https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit?id=4bacc9c9234c7c8eec44f5ed4e960d9f96fa0f01), so hopefully this will be fixed in v4.2.
